File: week-1/lecture/list_primes.py
"""List Primes example.

This file belongs to the introductory lecture examples on recursion, searching, and simple sorting. It uses get_primes, sieve_of_eratosthenes to illustrate the main idea with a small runnable example.
"""

def get_primes(n):
    """Return every prime number from 2 up to the given limit using trial division.

    Args:
        n (int): The n input used by this function.

    Returns:
        list: The list produced by the algorithm.
    """
    primes = []
    for num in range(2, n+1):
        is_prime = True
        for factor in range(2, num):
            if num % factor == 0:
                is_prime = False
                break
        if is_prime:
            primes.append(num)
    return primes

def sieve_of_eratosthenes(n):
    # Create a boolean array "prime[0..n]" and initialize
    # all entries it as true. A value in prime[i] will
    # finally be false if it is Not a prime, else true.
    """Return every prime number from 2 up to the given limit using the sieve method.

    Args:
        n (int): The n input used by this function.

    Returns:
        list: The list produced by the algorithm.
    """
    prime = [True for i in range(n+1)]
    p = 2
    while (p * p <= n):
        # If prime[p] is not changed, then it is a prime
        if (prime[p] == True):
            # Update all multiples of p
            for i in range(p * 2, n+1, p):
                prime[i] = False
        p += 1
    # Return all prime numbers
    primes = []
    for p in range(2, n+1):
        if prime[p]:
            primes.append(p)
    return primes


import timeit

n = 10000

result = timeit.timeit(stmt='get_primes(n)', globals=globals(), number=1)
print(f"Execution time is {result} seconds")

result = timeit.timeit(stmt='sieve_of_eratosthenes(n)', globals=globals(), number=1)
print(f"Execution time is {result} seconds")

File: week-1/lecture/list_primes.cpp
/**
 * @file list_primes.cpp
 * @brief C++ version of the prime-listing and timing example.
 */

#include <chrono>
#include <iostream>
#include <vector>

using namespace std;
using Clock = chrono::high_resolution_clock;

/**
 * @brief Return all prime numbers from 2 up to n using trial division.
 * @param n int Upper limit of the search. The function checks every integer from 2 to n.
 * @return vector<int> Prime numbers found up to and including n.
 */
vector<int> get_primes(int n) {
    vector<int> primes;

    for (int num = 2; num <= n; ++num) {
        bool is_prime = true;
        for (int factor = 2; factor * factor <= num; ++factor) {
            if (num % factor == 0) {
                is_prime = false;
                break;
            }
        }
        if (is_prime) {
            primes.push_back(num);
        }
    }

    return primes;
}

/**
 * @brief Return all prime numbers from 2 up to n using the sieve of Eratosthenes.
 * @param n int Upper limit of the search. The function marks multiples of smaller primes.
 * @return vector<int> Prime numbers found up to and including n.
 */
vector<int> sieve_of_eratosthenes(int n) {
    vector<bool> prime(n + 1, true);
    if (n >= 0) {
        prime[0] = false;
    }
    if (n >= 1) {
        prime[1] = false;
    }

    for (int p = 2; p * p <= n; ++p) {
        if (prime[p]) {
            for (int multiple = p * 2; multiple <= n; multiple += p) {
                prime[multiple] = false;
            }
        }
    }

    vector<int> primes;
    for (int value = 2; value <= n; ++value) {
        if (prime[value]) {
            primes.push_back(value);
        }
    }
    return primes;
}

/**
 * @brief Measure and print the runtime of the two prime-generation methods.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    int n = 10000;

    auto trial_start = Clock::now();
    vector<int> trial_primes = get_primes(n);
    auto trial_end = Clock::now();

    auto sieve_start = Clock::now();
    vector<int> sieve_primes = sieve_of_eratosthenes(n);
    auto sieve_end = Clock::now();

    chrono::duration<double> trial_time = trial_end - trial_start;
    chrono::duration<double> sieve_time = sieve_end - sieve_start;

    cout << "Trial division found " << trial_primes.size() << " primes in " << trial_time.count() << " seconds\n";
    cout << "Sieve found " << sieve_primes.size() << " primes in " << sieve_time.count() << " seconds\n";
    return 0;
}
